### PR TITLE
Call emitInitiateChange

### LINKED ISSUE #71: Call `emitInitiateChange` function by validator
(related to https://github.com/poanetwork/parity-ethereum/issues/49)

I've missed one detail:

There can be a case when the validator set is not changed at the beginning of new staking epoch (due to delegators did nothing during the finished staking epoch).

The `getPendingValidators()` getter will return the same set in that case, and as a result, the `finalizeChange()` function won't be called by Parity (because the set wasn't changed).

But we need to call `finalizeChange()` anyway when the new staking epoch begins.

Can we use the `changeRequestCount()` getter instead of `getPendingValidators()`? Its value is incremented each time when the set of validators is changed or when the new staking epoch begins.

### PR DESCRIPTION
**Pull request by DemiMarie**
Closes #71.

### CODE REVIEW

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 1]
```diff
@@ -291,10 +291,31 @@ impl ValidatorSet for ValidatorSafeContract {
 
 	fn on_epoch_begin(&self, _first: bool, _header: &Header, caller: &mut SystemCall) -> Result<(), ::error::Error> {
 		let data = validator_set::functions::finalize_change::encode_input();
-		caller(self.contract_address, data)
-			.map(|_| ())
+		let () = caller(self.contract_address, data)
+			.map(drop)
 			.map_err(::engines::EngineError::FailedSystemCall)
-			.map_err(Into::into)
+			.map_err(::error::Error::from)?;
```
[the remark this answers is not in the document]
Comment (DemiMarie) [reply ethcore/src/engines/validator_set/safe_contract.rs]: > I'm confused: Isn't `on_epoch_begin` only called after an `InitiateChange` event has been found? (See [this comment](https://github.com/poanetwork/parity-ethereum/issues/71#issuecomment-457167061) for my attempt to track this through the code.) [The doc comment](https://github.com/poanetwork/pos-contracts/blob/11196e234594da2000a2c744dc9e27961b27aa53/contracts/abstracts/ValidatorSetBase.sol#L25) also reads like `finalizeChange` is meant to be called _after_ `InitiateChange`.

You are correct.  This really needs to go in `ValidatorSet::on_new_block`, which doesn’t currently exist.  Should I create it, or drop the code in `AuthorityRound::on_new_block`?

[the remark this answers is not in the document]
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: I'd put it in `AuthorityRound::on_new_block` for now; it doesn't need anything from `SafeContract`, does it? Not sure whether that would conceptually be the right place for it.

[the remark this answers is not in the document]
Comment (DemiMarie) [reply ethcore/src/engines/validator_set/safe_contract.rs]: @afck It needs `ValidatorSafeContract::contract_address`.

[the remark this answers is not in the document]
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: You're right; probably `ValidatorSet::on_new_block` would make sense.

[the remark this answers is not in the document]
Comment (DemiMarie) [reply ethcore/src/engines/validator_set/safe_contract.rs]: @afck @varasev can `emitInitateChangeCallable` return `false` for all nodes other than the currently selected validator?  That would be nice.

[the remark this answers is not in the document]
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: With #75 we should be able to make sure that calls always apply to the block we're currently authoring, i.e. only if we're selected.

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 2]
```diff
@@ -32,10 +32,11 @@ use rlp::{Rlp, RlpStream};
 use std::sync::{Weak, Arc};
 use super::{SystemCall, ValidatorSet};
 use super::simple_list::SimpleList;
+use super::super::authority_round::util::BoundContract;
 use unexpected::Mismatch;
 use ethabi::FunctionOutputDecoder;
 
-use_contract!(validator_set, "res/contracts/validator_set.json");
+use_contract!(validator_set, "res/contracts/ValidatorSetAuRa.json");
```
Comment (afck) [inline ethcore/src/engines/validator_set/safe_contract.rs]: Aren't the other files in `res/contracts` all lowercase? I think we should try to keep it consistent.

File: ethcore/src/engines/authority_round/mod.rs [hunk 3]
```diff
@@ -49,7 +49,9 @@ use unexpected::{Mismatch, OutOfBounds};
 
 mod finality;
 mod randomness;
-mod util;
+pub(crate) mod util;
+
+use_contract!(validator_set, "res/contracts/ValidatorSetAuRa.json");
```
Comment (afck) [inline ethcore/src/engines/authority_round/mod.rs]: This doesn't seem to be used anywhere in this file?

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 2 above)
Comment (DemiMarie) [reply ethcore/src/engines/validator_set/safe_contract.rs]: I agree, with one catch: naming the files the same as in `pos-contracts` makes their automatic generation in bulk slightly easier.

File: ethcore/src/engines/authority_round/mod.rs (hunk 3 above)
Comment (DemiMarie) [reply ethcore/src/engines/authority_round/mod.rs]: Deleted.

File: ethcore/src/engines/validator_set/safe_contract.rs (hunk 2 above)
Comment (afck) [reply ethcore/src/engines/validator_set/safe_contract.rs]: I know. :pensive: But when working on Parity, I'd give Parity's conventions precedence over our own.

**Pull request closed without merging**